rfcs/2022-02-11-responsive-and-adaptive-layout-strategy.md

```diff
@@ -0,0 +1,414 @@
+
+### 1. Components that can handle Responsiveness Internally
+
+- Most of the components that don't deal with layouts shall handle responsiveness internally. For eg: Button, Modal, Avatar, Select, etc.
+
+- The properties that affects responsiveness are:
+
+  - `width`
+  - `height`
+  - `font-size`
+  - `padding`
+  - `margin`
+  - `position`
+  - `top`
+  - `left`
+  - `right`
+  - `bottom`
+
+- While a lot of [well known design systems gives flexibility](#approaches-taken-by-well-known-design-systems) on every component for above properties(becuse they might have taken a call to give flexibility to the consumers) to accept responsive values, I can't think of a use case where we'll have to provide similar functionality(because we want to strike a balance of flexibility and constraints, if you go read further we do have layout componments which will give this flexibility). Majority of the cases that I could think of wouldn't be any different even if the consumers define the values for them explicitly. So seems like we'll be adding unnecessary learning curve to the component's API.
+
+  - For eg: all the props(ones that are affected by responsive) for a component will then accept either single value or array/object of values for different screen sizes which is confusing and might lead to a lot of confusion at the API surface.
+
+    ```jsx
+    <Text fontSize="100" />
+
+    <Text fontSize={{ small: 100, medium: 100, large: 100, xl: 400 }} />
+    ```
+
+- Hence the idea here is that all these components except the Layout components will handle responsiveness based on predefined breakpoints internally without the consumers explicitly defining anything for them.
+
+### 2. Components that can give some flexibility to handle Responsiveness to consumers
+
+- While most of the components can handle responsiveness internally there are use cases for components where we would need to give flexibility to accept responsive values for certain props. For eg: Layout Components that handles the overall layout of the page might change the layout on different screen sizes based on the available space.
+  <img alt="Adaptive Layout with flex direction" src="./images/responsive-adaptive-layout/adaptive-layout-with-flex-direction.png">
+- The idea here is to have certain components that deal with Layout of the page expose some flexibility to consumers to tweak if they want the layout to look different on different screen sizes
+
+### 3. Components for Adaptive layout(rendering components conditionally)
+
+- After handling responsiveness internally and also giving flexibility for few scenarios there are still few complex cases that we need to handle:
+
+  - There are cases where we might need to render a completely different variation of a component based on the screen size and user experience. For eg: A Search Field component might be different for larger screen sizes and smaller screen sizes.
```

Comment:
> A Search Field component might be different for larger screen sizes and smaller screen sizes.

I think the consideration here should not be just about the "size" but also the device and how users interact with the component while optimising for the best user experience. 

We usually see a dropdown on desktop becoming a modal on mobile device to improve the overall UX of the component since mobile devices has a very small touch area thus trying to select a floating dropdown can be cumbersome. 

Also on another note:
Some components can be hybrid means they can be "Responsively Adaptive" 

Take an example of a Select component where the Select's input can be fluid, means it will stretch based on screensize but will also have the functionality to change the collapsible dropdown to modal on mobile devices.

Ideally we should have a list of all the component which needs adaptive strategy, and ones which needs both strategies.

Reply:
@anuraghazra that's the summary of the entire RFC, not sure if you meant to add something on this line?

Reply:
Yup got it, Was reviewing async, didn't read the entire thing while commenting this. 

But second paragraph about "Responsively Adaptive" 
We are going to be adopting hybrid strategy if needed right?

Reply:
yes that's what we have [mentioned here](https://github.com/razorpay/blade/pull/416/files#diff-f4d103560de689c11de7f41e8dbd9106c41e45f89247d0d2508f9150040c463dR85)